### PR TITLE
Import Bad Request Duplicate Files

### PR DESCRIPTION
**Pull request by mikaelweave**
## Description
Changes `$import` behavior to immediately return with bad request if duplicate files are submitted.

## Related issues
AB#1117694

## Testing
Included new unit test and E2E test.
Tested locally.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

**Review by LTA-Thinking — DISMISSED** [review]

**Review by SergeyGaluzo — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportOrchestratorJob.cs [hunk 1]
```diff
@@ -412,7 +412,7 @@ private async Task<IList<long>> EnqueueProcessingJobsAsync(IEnumerable<InputReso
             {
                 const string message = "Duplicate file detected in list of files to import.";
                 _logger.LogJobError(ex, orchestratorInfo, message);
-                throw new JobExecutionException(message, ex);
+                throw new ImportProcessingException(message, ex);
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportOrchestratorJob.cs:415]: > throw new ImportProcessingException(message, ex); [](http://example.com/codeflow?start=15&length=50)

What is the reason to register $import? Why not to return bad request from import controller before registration?

**Comment by SergeyGaluzo** [comment]
>             catch (SqlException ex) when (ex.Number == 2627)

I think it is a wrong place to check whether input list is unique. We should return 400 from import controller without registration.

---
Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportOrchestratorJob.cs:411 in a78b1b7. [](commit_id = a78b1b7102b1eaa12467e588741cbbb10670dd6b, deletion_comment = False)

**Comment by mikaelweave** [comment]
> > ```
> >         catch (SqlException ex) when (ex.Number == 2627)
> > ```
> 
> I think it is a wrong place to check whether input list is unique. We should return 400 from import controller without registration.
> 
> Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/ImportOrchestratorJob.cs:411 in [a78b1b7](https://github.com/microsoft/fhir-server/commit/a78b1b7102b1eaa12467e588741cbbb10670dd6b). [](commit_id = [a78b1b7](https://github.com/microsoft/fhir-server/commit/a78b1b7102b1eaa12467e588741cbbb10670dd6b), deletion_comment = False)

Makes sense - should be a bad request from the start. I will add this to the controller and remove the check from the job itself.

**Comment by SergeyGaluzo** [comment]
I think it is fixed by https://github.com/microsoft/fhir-server/pull/3796

**Comment by mikaelweave** [comment]
> I think it is fixed by #3796

In general, yes you are correct. However I also implemented a check at the `POST /$import` level as suggested in the PR comments. This was not fixed by #3796. 😄

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs [hunk 2]
```diff
@@ -198,6 +198,14 @@ private void ValidateImportRequestConfiguration(ImportRequest importData)
                 throw new RequestNotValidException(string.Format(Resources.ImportRequestValueNotValid, nameof(input)));
             }
 
+            var duplicateInputUrls = input.GroupBy(item => item.Url).Where(group => group.Count() > 1).Select(group => group.Key);
+
+            if (duplicateInputUrls.Any())
+            {
+                var duplicateUrlString = string.Join(", ", duplicateInputUrls);
+                throw new RequestNotValidException(string.Format(Resources.ImportRequestDuplicateInputFiles, duplicateUrlString));
+            }
```
Comment (SergeyGaluzo) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs:207]: Please remove dup exception logic from orchestrator.

Comment (mikaelweave) [reply src/Microsoft.Health.Fhir.Shared.Api/Controllers/ImportController.cs:207]: done thx for calling that out

**Review by SergeyGaluzo — APPROVED** [review]
:shipit:

**Review by LTA-Thinking — APPROVED** [review]

**Pull request merged**